MCOL-4617 Move in-to-exists predicate creation and injection into the engine.

tntnatbry opened the pull request:
We earlier leveraged the server functionality provided by

Item_in_subselect::create_in_to_exists_cond and
Item_in_subselect::inject_in_to_exists_cond

to create and inject the in-to-exists predicate into an IN
subquery's JOIN struct. With this patch, we leave the IN subquery's
JOIN unaltered and instead directly perform this predicate creation
and injection into ColumnStore's select execution plan.

tntnatbry committed:
MCOL-4617 Move in-to-exists predicate creation and injection into the engine.

We earlier leveraged the server functionality provided by

Item_in_subselect::create_in_to_exists_cond and
Item_in_subselect::inject_in_to_exists_cond

to create and inject the in-to-exists predicate into an IN
subquery's JOIN struct. With this patch, we leave the IN subquery's
JOIN unaltered and instead directly perform this predicate creation
and injection into ColumnStore's select execution plan.

davidhall586 reviewed (commented):
In buildInToExistsFilter(), It calls buildEqualityPredicate(). It's possible for buildEqualityPredicate() to find an error and set gwip->fatalParseError. Should we check for this condition in the caller?

I may be missing something, but the code at line 1804 appears duplicated at line 1821. I realize this is old code, but does it do anything useful the second time?

drrtuy reviewed (commented):
Could you ellaborate on why there is no examples with correlated subquery in the test file?

drrtuy commented on dbcon/mysql/ha_in_sub.cpp:
```diff
@@ -125,10 +137,9 @@ execplan::ParseTree* InSub::transform()
 
     fSub = (Item_subselect*)(fFunc->arguments()[1]);
-    idbassert(fSub && fFunc);
+    idbassert(fSub);
```
This is a poor man way to send the related fFunc into the log so plz left it intact or suggest your way to relay this information.

tntnatbry replied:
This is not the case. `idbassert(fSub && fFunc)` will print the string `"fsub && fFunc"` in the log, rather than the actual values of the arguments. Since we already check for `fFunc` at the start of the function, we don't need to add an assertion check on `fFunc` here again.

drrtuy replied:
Agree.

drrtuy commented on dbcon/mysql/ha_mcs_impl_if.h:
```diff
@@ -177,7 +183,8 @@ struct gp_walk_info
         cs_vtable_impossible_where_on_union(false),
-        isGroupByHandler(false)
+        isGroupByHandler(false),
+        inSubQueryLHS(0)
```
There should be a default for inSubQueryLHSItem

drrtuy replied:
Plz use nullptr instead of 0 constant.

tntnatbry replied:
Done.

drrtuy commented on dbcon/mysql/ha_mcs_execplan.cpp:
```diff
@@ -1677,6 +1677,188 @@ bool buildRowColumnFilter(gp_walk_info* gwip, RowColumn* rhs, RowColumn* lhs, It
+    execplan::ReturnedColumn* rhs,
+    gp_walk_info* gwip,
+    boost::shared_ptr<Operator>& sop,
+    Item_func::Functype funcType,
```
I don't see that we change this argument so it should be const.

tntnatbry replied:
Done.

drrtuy commented on dbcon/mysql/ha_mcs_execplan.cpp:
```diff
@@ -6849,6 +6868,115 @@ int processLimitAndOffset(
+
+        int i = 0;
+
+        ParseTree* rowFilter = 0;
```
nullptr plz.

tntnatbry replied:
Done.

drrtuy commented on dbcon/mysql/ha_mcs_execplan.cpp:
```diff
@@ -6849,6 +6868,115 @@ int processLimitAndOffset(
+        }
+
+        // Delete the duplicate copy of the returned cols
+        auto iter = gwi.additionalRetCols.begin() + additionalRetColsBefore;
```
Plz use std::advance here.

tntnatbry replied:
Done.

drrtuy commented on dbcon/mysql/ha_mcs_execplan.cpp:
```diff
@@ -1677,6 +1677,188 @@ bool buildRowColumnFilter(gp_walk_info* gwip, RowColumn* rhs, RowColumn* lhs, It
+    // column list, so the materialized view have the complete projection list.
+    // e.g. tout.c1 in (select tin.c1 from tin where tin.c2=tout.c2);
+    // the projetion list of subquery will have tin.c1, tin.c2.
+    ReturnedColumn* correlatedCol = NULL;
```
plz use nullptr.

tntnatbry replied:
Done.

drrtuy commented:
I don't see any serious functional issue so plz address the comments or comment back so I can proceed.

drrtuy commented:
You should also rebase the patch on top of the current develop so that MTR failure goes away.

tntnatbry commented:
@dhall-MariaDB I have addressed both your comments, thanks for highlighting them. Please review again.

@drrtuy I have added test case for correlated subquery under the comment `### Correlated IN subquery`, as well, also addressed most of your additional comments. Please review again.
